### PR TITLE
Fix let pun locations

### PR DESCRIPTION
**Pull request by ccasin**
This PR fixes an issue with locations in let puns.  The pun:
```ocaml
let%bind x in ...
```
desugars into
```ocaml
let%bind x = x in ...
```
The problem is that neither the pattern nor the expression are given a ghost location. This breaks `ppxlib`'s invariant checks, which mandate that sibling ast notes may not have overlapping (non-ghost) locations.

So, to support let punning, we must make one of the ghost.

Which one?  I made the pattern's location ghost, for two reasons:
* The reason ppxlib wants one of these two locations to be ghost is so that merlin knows what is at the location - a pattern or an expression.  What might someone do on this punned variable in merlin?  Autocompleting seems reasonable, and that's a thing you do on expressions.  Destructing seems less reasonable, and that's a thing you do on patterns. So we should want the "real" thing to be the expression, for merlin.
* For other things that can be punned (like record field patterns and expressions) the thing on the left in the expanded version is given a ghost location.  Though there the thing on the left is a label and here it is a pattern, so this doesn't feel like much of a reason.

I'm not sure what kind of test should be added to the compiler for this, if any, but I did test locally that ppxlib no longer gets angry about a punned let binding in a situation where it did before.

### CODE REVIEW

**Review by ncik-roberts — COMMENTED** [review]
I appreciate the alacrity.

**Theme 1: ghostiness**

I think the expression should be ghost and the pattern should not be ghost.

I think this because:
  * Autocomplete still works even if the expression is ghost. (I tested this with a patched version of merlin.)
  * The type of the pattern is, IMO, a nicer thing to print in a merlin type-enclosing query.
  * Merlin prints the type of the pattern today.

I snuck an opinion in as the second bullet point. But I stand by it. `let%...` and `let*` are most often used in monadic code [citation needed], and so when I ask for the type of `x` in `let* x in ...`, it's redundant to be told that this is a monadic use of `x`. I know that. I'd rather just be told the type parameter.

**Theme 2: On ocaml/ocaml**

We should make this change in ocaml/ocaml as well. However, ppxlib's check is disabled by default. So that check is less clearly the main motivating reason for this change. Maybe the real reason is something like "merlin is working now, but only by accident". And the existence of ppxlib's check doesn't hurt &mdash; we should do simple things to make it happy.

**Comment by ccasin** [comment]
> I think the expression should be ghost and the pattern should not be ghost.
> 
> ...

Funny, my instinct was to prefer printing the type of the expression, as you might wonder what monad you are in when reading someone else's code.  Admittedly, this is usually clear from relatively nearby context, but it doesn't have to be.

But, I'm quite happy either way, so I've changed it.

> We should see if there’s appetite for this change in ocaml/ocaml as well. However, ppxlib's check is disabled by default. So that check is less clearly the main motivating reason for this change. Maybe the real reason is something like "merlin is working now, but only by accident". And the existence of ppxlib's check doesn't hurt — we should do simple things to make it happy.

Agreed - I will make a PR upstream today or over the weekend.  Let's not hold this up waiting for upstream approval though - it's wanted internally, and we can integrate upstream feedback in a future PR.

File: ocaml/parsing/parser.mly
Comment (ccasin) [reply ocaml/parsing/parser.mly]: Thanks, fixed.  This was sloppy, apolgies.

**Review by ncik-roberts — APPROVED** [review]

**Comment by ncik-roberts** [comment]
When we upstream this, we should make sure we also upstream #2612 as part of it.

**Pull request merged**
